File: 2020/php/src/Aoc/Days/Day8.php
<?php


namespace Ppx17\Aoc2020\Aoc\Days;


use Illuminate\Support\Collection;

class Day8 extends AbstractDay
{
    private Collection $instructions;

    public function dayNumber(): int
    {
        return 8;
    }

    public function setUp(): void
    {
        $this->instructions = collect($this->getInputLines())
            ->map(fn($x) => explode(' ', $x));
    }

    public function part1(): string
    {
        return $this->run($this->instructions)[1];
    }

    private function run(Collection $instructions): array
    {
        $seen = [];
        $acc = $ptr = 0;
        $instructionCount = $instructions->count();
        while (!isset($seen[$ptr])) {
            $seen[$ptr] = true;
            [$op, $param] = $instructions->get($ptr);
            switch($op) {
                case 'nop': $ptr++; break;
                case 'acc': $acc += $param; $ptr++; break;
                case 'jmp': $ptr += $param; break;
            }
            if ($ptr === $instructionCount) {
                return ['exit', $acc];
            }
        }
        return ['loop', $acc];
    }

    public function part2(): string
    {
        foreach ($this->instructions as $key => $instruction) {
            [$op, $param] = $instruction;
            if ($op === 'jmp' || $op === 'nop') {
                $copy = clone $this->instructions;
                $copy->put($key, [$op === 'jmp' ? 'nop' : 'jmp', $param]);
                [$mode, $acc] = $this->run($copy);
                if ($mode === 'exit') {
                    return $acc;
                }
            }
        }
    }
}

File: 2020/julia/src/Day8.jl
module Day8
include("Aoc.jl")
using .Aoc

instructions = map((x) -> (x[1], parse(Int64, x[2])), split.(Aoc.input_lines(8)))

function run(instructions)
    seen = Dict{Int64,Bool}()
    incount = length(instructions)
    acc = ptr = 0

    while !haskey(seen, ptr)
        seen[ptr] = true
        (op, param) = instructions[ptr + 1]
        if op == "nop"
            ptr += 1
        elseif op == "acc"
            acc += param
            ptr += 1
        elseif op == "jmp"
            ptr += param
        end
        if ptr == incount
            return (true, acc)
        end
    end
    (false, acc)
end

part1() = run(instructions)[2]

function part2()
    for (idx, (op, param)) in enumerate(instructions)
        op == "acc" && continue
        instr = copy(instructions)
        instr[idx] = (op == "jmp" ? "nop" : "jmp", param)
        (exit, acc) = run(instr)
        exit && return acc
    end
end
end
